### PR TITLE
Face off: Virtual DOM vs Glimmer vs Incremental DOM article

### PR DESCRIPTION
**Pull request by sebadoom**
Draft. RFC. Missing: benchmarks.

face-off react js virtual dom vs incremental dom vs ember js glimmer

### CODE REVIEW

File: _posts/2015-11-16-face-off-virtual-dom-vs-incremental-dom-vs-glimmer.markdown [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+- glimmer
+- benchmark
+---
+
+In this post we will explore three technologies to build dynamic DOMs. We will also run benchmarks and find out which one is faster. We will also share with you why we choose one of them for our projects. Read on!
+
+-----
+
+## Introduction
+There are many [Document Object Model (DOM)](https://developer.mozilla.org/en-US/docs/Web/API/Document_Object_Model) manipulation frameworks and libraries. Among these, three have been the center of the attention due to their focus on performance: [React.js](https://facebook.github.io/react/), [Ember.js](http://emberjs.com/) and, more recently, [Incremental DOM](https://github.com/google/incremental-dom). While React and Ember handle much more than just DOM building/update, Incremental DOM focuses only on building DOM trees and allowing dynamic updates. We will now explore these libraries and find out which one is faster.
```
Comment (mgonto) [inline _posts/2015-11-16-face-off-virtual-dom-vs-incremental-dom-vs-glimmer.markdown]: Add in the introduction some more explanation about WHY this frameworks started to show up. DOM manipulations are slow, repaints are slow, doing more operations than needed are slow.

Give as an example

show an element
then hide
then show
then hide

That runs 4 diff operations which require repaint. If you do Virtual Dom it'll only run one in the end.

Also, Mention some more about the techs they use (Virtual DOM vs Glimmer) not just the techlogies

File: _posts/2015-11-16-face-off-virtual-dom-vs-incremental-dom-vs-glimmer.markdown [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+Incremental DOM does not favor any particular template engine. However, being a Google sponsored project, a [Closure templates](https://github.com/google/closure-templates/) backend is being developed. Incremental DOM can also be used with [superviews.js](https://github.com/davidjamesstone/superviews.js), [starplate](https://github.com/littlstar/starplate) and even [JSX](https://github.com/babel-plugins/babel-plugin-incremental-dom).
+
+## React.js' Virtual DOM
+*Virtual DOM* is the name React developers gave to their DOM manipulation engine. *Virtual DOM* provides a series of Javascript calls that tell the library how to build a DOM tree and how to update it when data bound to it changes. The central piece of Virtual DOM is its smart diffing algorithm: once the differences in the model have been mapped to the in-memory copy of the DOM, the algorithm finds the minimum number of operations required to update the real DOM.
```
Comment (mgonto) [inline _posts/2015-11-16-face-off-virtual-dom-vs-incremental-dom-vs-glimmer.markdown]: This isn't 100% clear. I'd rephrase. Start telling that it creates an In memory copy of dom, that it does operations there, then it does the diff on every Javascript turn end. Focus on this in memory tree and how it works

File: _posts/2015-11-16-face-off-virtual-dom-vs-incremental-dom-vs-glimmer.markdown [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+Glimmer differentiates between **static** and **dynamic** components, thus reducing the number of elements that need to be checked when looking for changes. This differentiation can be achieved thanks to the expressiveness of Handlerbar's templates.
+
+Another key difference between Glimmer and other solutions lies in the way nodes are stored and compared. Glimmer stores nodes as simple stream-like objects (that is, simple queues of values) rather than full-fledged DOM-like nodes. To find out whether a real DOM node needs updating, the final value of a Glimmer node is compared to the last known real DOM value. If the value has not changed, no further actions are taken. The final value of a Glimmer node is simply the result of flushing the stream of values stored in it. Events and other transient operations work on the values stored in Glimmer's stream-like nodes, resulting in a single batch of updates to the real DOM (rather than many) once all operations are complete.
```
Comment (mgonto) [inline _posts/2015-11-16-face-off-virtual-dom-vs-incremental-dom-vs-glimmer.markdown]: This is nice but too complex. Can we simplify this explanation?

File: _posts/2015-11-16-face-off-virtual-dom-vs-incremental-dom-vs-glimmer.markdown [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+- Differentiation between static and dynamic elements
+- 100% compatible with Ember's API (you get the benefits without major updates to your existing code)
+- Lightweight in-memory representation of the DOM
+
+#### Cons
+- Meant to be used only in Ember
+- Only one frontend available
+
+## Incremental DOM
+Incremental DOM tries to bring a simpler approach to the table than the alternatives: rather than keeping a full in-memory representation of the DOM, or keeping a tree of lightweight elements, Incremental DOM uses the **real DOM** to find differences when data changes. You might be asking yourself why, if this is simpler, it hasn't been the solution picked by other libraries all along. Simple: it results in a **tradeoff between speed and memory**. Incremental DOM, by removing the additional copy of the DOM, results in **reduced memory usage**. In practice this also results in **reduced speed** while looking for differences. The reduced memory usage is key for mobile or other memory constrained devices. Read [our article on Incremental DOM](https://auth0.com/blog/2015/10/23/incremental-dom/) for more information on this.
```
Comment (mgonto) [inline _posts/2015-11-16-face-off-virtual-dom-vs-incremental-dom-vs-glimmer.markdown]: This one is erally clear!

File: _posts/2015-11-16-face-off-virtual-dom-vs-incremental-dom-vs-glimmer.markdown [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+- react
+- dom
+- virtual dom
+- incremental dom
+- javascript
+- glimmer
+- benchmark
+---
+
+In this post we will explore three technologies to build dynamic DOMs. We will also run benchmarks and find out which one is faster. We will also share with you why we choose one of them for our projects. Read on!
```
Comment (diegopoza) [inline _posts/2015-11-16-face-off-virtual-dom-vs-incremental-dom-vs-glimmer.markdown]: You are using "We will also..." in two contiguous sentences, can we use another phrase, like "Additionally, we will share ..."

File: _posts/2015-11-16-face-off-virtual-dom-vs-incremental-dom-vs-glimmer.markdown (hunk 1 above)
Comment (sebadoom) [inline _posts/2015-11-16-face-off-virtual-dom-vs-incremental-dom-vs-glimmer.markdown]: Some more about the technologies? I don't quite get what you mean in that last sentence. Could you clarify?

Comment (sebadoom) [inline _posts/2015-11-16-face-off-virtual-dom-vs-incremental-dom-vs-glimmer.markdown]: I don't think these frameworks were originally designed because DOM updates are slow. I think they were designed to aid in data binding. Speed became a problem later, when lots of bindings were in place for big sites, I think. I'll find a way to mention concerns about speed though.

File: _posts/2015-11-16-face-off-virtual-dom-vs-incremental-dom-vs-glimmer.markdown (hunk 2 above)
Comment (sebadoom) [inline _posts/2015-11-16-face-off-virtual-dom-vs-incremental-dom-vs-glimmer.markdown]: I will add a couple of clarifications here and there. It describes exactly what you mentioned. Let's see if the clarifications make it is easier to understand.

File: _posts/2015-11-16-face-off-virtual-dom-vs-incremental-dom-vs-glimmer.markdown (hunk 3 above)
Comment (sebadoom) [inline _posts/2015-11-16-face-off-virtual-dom-vs-incremental-dom-vs-glimmer.markdown]: I will remove some info to make it simpler. We do need to mention that values are stored in the in-memory DOM tree to compare for differences, otherwise the diagram will not make sense.

**Comment by sebadoom** [comment]
Implemented latest comments. Still missing: benchmarks.
face-off react js virtual dom vs incremental dom vs ember js glimmer 1

**Comment by mgonto** [comment]
Sold. ETA for the diagrams? Could we merge it tomorrow?

**Comment by sebadoom** [comment]
Should be done today (thursday). All benchmarks are running. We still need to analyze the data and make cute charts.

**Comment by sebadoom** [comment]
Full article. Missing: icon (need help!), SEO checklist.
face-off react js virtual dom vs incremental dom vs ember js glimmer
